core/blockchain.go

```diff
@@ -376,6 +379,10 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *par
 	bc.prefetcher = newStatePrefetcher(chainConfig, bc.hc)
 	bc.processor = NewStateProcessor(chainConfig, bc.hc)
 
+	if cacheConfig.TrieTimeLimitRandomOffset > 0 {
+		bc.gcproc = time.Duration(rand.Int64N(int64(cacheConfig.TrieTimeLimitRandomOffset)))
```

Comment:
need to also set whenever setting gcproc to 0

Reply:
specifically:
* FlushTrieDB
* writeBlockchainWithState

Reply:
I don't think we need to randomly offset the gcproc after each commit. Spreading the commits once should make it more predictable (if it spreads nicely at start it shouldn't drift towards time collision soon) and "TrieTimeLimit" will be still working the same as before.

Or do you think that it be better to random offset each time? so if we are unlucky with the first distribution, each next one will be again randomized? 
If you think it's better to randomize each time, I'll implement it a bit differently, so as to keep gcproc intact so the logged value of processing time since last commit will be accurate.

Reply:
I think we shoud randomize each time, to make sure we never enter a steady-state of many synchronizations.
One important point is that we should randomize only each time that we flush (randomize only on startup has some potential for weird steady state, randomize on every time we measure gcpro - wil always flush when random selected will be close to zero).
I accept your point about wanting to keep gcproc's meaning.
Maybe best would be to add a separate "randomOffset" field that is done separately.

Reply:
done, added `bc.gcprocRandOffset` which is added to `gcproc` during check against `flushInterval` so the full trie flush still happens no later then `trie-time-limit`, but can happen up to `trie-time-limit-random-offset` earlier. The reasoning behind it is that `trie-time-limit` will still be the maximum time we want to spend re-executing in case of unclean shutdown.